contrib/dockerfile_build.bzl

```diff
@@ -59,63 +70,165 @@ def _impl(repository_ctx):
+            dockerfile_path.dirname,
+        ])
+
+        build_result = repository_ctx.execute(command)
+        if build_result.return_code:
+            fail("Image build command failed: {} ({})".format(
+                build_result.stderr,
+                " ".join(command),
+            ))
+
+        save_result = repository_ctx.execute([
+            docker_path,
+            "save",
+            "-o",
+            "image/" + _OUTPUT_IMAGE_TAR,
+            img_name,
+        ])
+        if save_result.return_code:
+            fail("docker save command failed for image {}: {}".format(
+                img_name,
+                save_result.stderr,
+            ))
+
+    elif repository_ctx.attr.driver == "kaniko":
+        # Additional kaniko options.
+        kaniko_flags = [
+            # Path to Dockerfile within the build context.
+            "--dockerfile=Dockerfile",
+            # Image tag.
+            "-d",
+            img_name,
+            # Strip timestamps.
+            "--reproducible",
+            "--no-push",
+        ]
+
+        template = repository_ctx.path(Label("@io_bazel_rules_docker//contrib:kaniko_run_and_extract.sh.tpl"))
+        repository_ctx.template(
+            "kaniko_run_and_extract.sh",
+            template,
+            {
+                "%{build_context_dir}": str(dockerfile_path.dirname),
+                "%{docker_path}": docker_path,
+                "%{extract_file}": "/%s" % _OUTPUT_IMAGE_TAR,
+                "%{image_path}": repository_ctx.attr.kaniko_image_path,
+                "%{kaniko_flags}": " ".join(kaniko_flags),
+                "%{kaniko_workspace}": _KANIKO_WORKSPACE_DIR,
+                "%{output}": "%s/%s" % (repository_ctx.path("image"), _OUTPUT_IMAGE_TAR),
+            },
+            True,
+        )
+
+        build_result = repository_ctx.execute(["./kaniko_run_and_extract.sh"])
+        if build_result.return_code:
+            fail("Image build command failed: {}".format(
+                build_result.stderr,
+            ))
+
+    else:
+        # Should not hit here.
+        fail("Driver `%s` is not supported." % repository_ctx.attr.driver)
+
+_dockerfile_image = repository_rule(
     attrs = {
         "build_args": attr.string_dict(
             doc = "A map of args to pass to the --build-arg option in the " +
-                  "`docker build` command.",
+                  "docker or kaniko build command.",
         ),
         "docker_path": attr.string(
-            doc = "The full path to the docker binary. If not specified, it " +
-                  "will be searched for in the path. If not available, " +
-                  "the rule build will fail.",
+            doc = "The full path to the docker binary.",
         ),
         "dockerfile": attr.label(
             allow_single_file = True,
             mandatory = True,
             doc = "The label for the Dockerfile to build the image from.",
         ),
+        "driver": attr.string(
```

Comment:
What's the default?
I see default below, make mandatory then

Reply:
Done.